crates/bevy_ecs/src/bundle.rs

```diff
@@ -1128,12 +1259,16 @@ impl Bundles {
             let mut component_ids = Vec::new();
             T::component_ids(components, storages, &mut |id| component_ids.push(id));
             let id = BundleId(bundle_infos.len());
+            let mut required_components = RequiredComponents::default();
+            T::register_required_components(components, storages, &mut required_components);
+            required_components.remove_bundle_components(&component_ids);
+            let required_components = required_components.0.drain().map(|(_, v)| v).collect();
```

Comment:
Draining mapping and recollecting feels wasteful. Why can't we just pass in required_components.values()?

Reply:
BundleInfo needs an owned collection of RequiredComponent, so `required_components.values()` doesn't work here because it is a borrowing iterator.

In theory we could just store the HashMap directly instead of converting it to a Vec, but:

1. With the addition of `require(construct(X))`, the map will have entries that could either be Construct or Immediate. These will then need to be split out into separate lists for BundleInfo.
2. The `values()` HashMap iterator is slower than iterating a Vec. Given that iterating the list will be in the much hotter path of "work done for every entity", doing the work here (once) to convert that to a Vec will almost certainly pay dividends in the majority of cases.

Reply:
Performance-wise this seems fine to me since it will be called only once per `Bundle`.

The only thing weird seems to be the `.drain()`. Why not `.into_iter()`?

Reply:
`into_iter()` seems like the better choice in this example. If you look at the documentation for `HashMap::drain()`, it says:

> "Clears the map, returning all key-value pairs as an iterator. Keeps the allocated memory for reuse." ([Source](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.drain))

It looks like `drain()` is best when you intend on re-using the `HashMap`, but in this case `required_components` gets shadowed then dropped at the end of this function.

```suggestion
            let required_components = required_components.0.into_iter().map(|(_, v)| v).collect();
```

Reply:
I've swapped to into_iter. That does make more sense.